File: packages/google_maps_places_autocomplete_widgets_native/ios/google_maps_places_autocomplete_widgets_native/Sources/google_maps_places_autocomplete_widgets_native/PlacesNativePlugin.swift
import Flutter
import GooglePlaces
import FirebaseAppCheck

public class PlacesNativePlugin: NSObject, FlutterPlugin, PlacesNativeApi {
  // Session token lives natively: created lazily, consumed by fetchPlace.
  private var sessionToken: GMSAutocompleteSessionToken?
  // Retained so the SDK's weakly-held provider reference stays alive.
  private static var tokenProvider: TokenProvider?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = PlacesNativePlugin()
    PlacesNativeApiSetup.setUp(binaryMessenger: registrar.messenger(), api: instance)
  }

  class TokenProvider: NSObject, GMSPlacesAppCheckTokenProvider {
    func fetchAppCheckToken() async throws -> String {
      return try await AppCheck.appCheck().token(forcingRefresh: false).token
    }
  }

  func initialize(apiKey: String, useAppCheck: Bool,
                  completion: @escaping (Result<Void, Error>) -> Void) {
    GMSPlacesClient.provideAPIKey(apiKey)
    if useAppCheck {
      let provider = TokenProvider()
      PlacesNativePlugin.tokenProvider = provider
      GMSPlacesClient.setAppCheckTokenProvider(provider)
    }
    completion(.success(()))
  }

  func fetchPredictions(query: String, includedTypes: [String],
                        countryCode: String?, languageCode: String?,
                        completion: @escaping (Result<[NativePrediction], Error>) -> Void) {
    if sessionToken == nil { sessionToken = GMSAutocompleteSessionToken() }
    let filter = GMSAutocompleteFilter()
    filter.types = includedTypes
    if let country = countryCode { filter.countries = [country] }
    let request = GMSAutocompleteRequest(query: query)
    request.filter = filter
    request.sessionToken = sessionToken
    GMSPlacesClient.shared().fetchAutocompleteSuggestions(from: request) { results, error in
      if let error = error { return completion(.failure(error)) }
      let predictions: [NativePrediction] = (results ?? []).compactMap { r in
        guard let s = r.placeSuggestion else { return nil }
        return NativePrediction(
            placeId: s.placeID,
            fullText: s.attributedFullText.string,
            primaryText: s.attributedPrimaryText.string,
            secondaryText: s.attributedSecondaryText?.string,
            types: s.types)
      }
      completion(.success(predictions))
    }
  }

  func fetchPlace(placeId: String, languageCode: String?,
                  completion: @escaping (Result<NativePlaceDetails, Error>) -> Void) {
    let properties = [
      GMSPlaceProperty.name, GMSPlaceProperty.formattedAddress,
      GMSPlaceProperty.addressComponents, GMSPlaceProperty.coordinate
    ].map { $0.rawValue }
    let request = GMSFetchPlaceRequest(
        placeID: placeId, placeProperties: properties, sessionToken: sessionToken)
    sessionToken = nil  // details call terminates the billing session
    GMSPlacesClient.shared().fetchPlace(with: request) { place, error in
      if let error = error { return completion(.failure(error)) }
      guard let place = place else {
        return completion(.failure(NSError(domain: "PlacesNative", code: 404)))
      }
      let components = (place.addressComponents ?? []).map { c in
        NativeAddressComponent(types: c.types, longText: c.name, shortText: c.shortName)
      }
      completion(.success(NativePlaceDetails(
          name: place.name,
          formattedAddress: place.formattedAddress,
          lat: place.coordinate.latitude,
          lng: place.coordinate.longitude,
          components: components)))
    }
  }
}

File: packages/google_maps_places_autocomplete_widgets_native/android/src/main/kotlin/com/timmaffett/google_maps_places_autocomplete_widgets_native/PlacesNativePlugin.kt
package com.timmaffett.google_maps_places_autocomplete_widgets_native

import android.content.Context
import com.google.android.libraries.places.api.Places
import com.google.android.libraries.places.api.model.AutocompleteSessionToken
import com.google.android.libraries.places.api.model.Place
import com.google.android.libraries.places.api.net.FetchPlaceRequest
import com.google.android.libraries.places.api.net.FindAutocompletePredictionsRequest
import com.google.android.libraries.places.api.net.PlacesClient
import com.google.android.libraries.places.api.auth.PlacesAppCheckTokenProvider
import com.google.common.util.concurrent.ListenableFuture
import com.google.common.util.concurrent.SettableFuture
import com.google.firebase.appcheck.FirebaseAppCheck
import io.flutter.embedding.engine.plugins.FlutterPlugin

class PlacesNativePlugin : FlutterPlugin, PlacesNativeApi {
  private lateinit var context: Context
  private var client: PlacesClient? = null
  // Session token lives natively: created lazily, consumed by fetchPlace.
  private var sessionToken: AutocompleteSessionToken? = null

  override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    context = binding.applicationContext
    PlacesNativeApi.setUp(binding.binaryMessenger, this)
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    PlacesNativeApi.setUp(binding.binaryMessenger, null)
  }

  override fun initialize(apiKey: String, useAppCheck: Boolean, callback: (Result<Unit>) -> Unit) {
    try {
      Places.initializeWithNewPlacesApiEnabled(context, apiKey)
      if (useAppCheck) {
        Places.setPlacesAppCheckTokenProvider(FirebaseTokenProvider())
      }
      client = Places.createClient(context)
      callback(Result.success(Unit))
    } catch (e: Exception) {
      callback(Result.failure(e))
    }
  }

  private class FirebaseTokenProvider : PlacesAppCheckTokenProvider {
    override fun fetchAppCheckToken(): ListenableFuture<String> {
      val future = SettableFuture.create<String>()
      FirebaseAppCheck.getInstance()
          .getAppCheckToken(false)
          .addOnSuccessListener { future.set(it.token) }
          .addOnFailureListener {
            // Surfaced at warning level: with enforcement on this is the
            // difference between working and rejected requests.
            android.util.Log.w("PlacesNativePlugin", "App Check token fetch failed", it)
            future.setException(it)
          }
      return future
    }
  }

  override fun fetchPredictions(
      query: String,
      includedTypes: List<String>,
      countryCode: String?,
      languageCode: String?,
      callback: (Result<List<NativePrediction>>) -> Unit
  ) {
    val c = client ?: return callback(Result.failure(IllegalStateException("Places not initialized")))
    if (sessionToken == null) sessionToken = AutocompleteSessionToken.newInstance()
    val builder = FindAutocompletePredictionsRequest.builder()
        .setQuery(query)
        .setSessionToken(sessionToken)
        .setTypesFilter(includedTypes)
    if (countryCode != null) builder.setCountries(listOf(countryCode))
    c.findAutocompletePredictions(builder.build())
        .addOnSuccessListener { response ->
          callback(Result.success(response.autocompletePredictions.map { p ->
            NativePrediction(
                placeId = p.placeId,
                fullText = p.getFullText(null).toString(),
                primaryText = p.getPrimaryText(null).toString(),
                secondaryText = p.getSecondaryText(null).toString(),
                types = p.types)
          }))
        }
        .addOnFailureListener { callback(Result.failure(it)) }
  }

  override fun fetchPlace(
      placeId: String,
      languageCode: String?,
      callback: (Result<NativePlaceDetails>) -> Unit
  ) {
    val c = client ?: return callback(Result.failure(IllegalStateException("Places not initialized")))
    val fields = listOf(
        Place.Field.DISPLAY_NAME, Place.Field.FORMATTED_ADDRESS,
        Place.Field.ADDRESS_COMPONENTS, Place.Field.LOCATION)
    val request = FetchPlaceRequest.builder(placeId, fields)
        .setSessionToken(sessionToken)
        .build()
    sessionToken = null  // details call terminates the billing session
    c.fetchPlace(request)
        .addOnSuccessListener { response ->
          val place = response.place
          callback(Result.success(NativePlaceDetails(
              name = place.displayName,
              formattedAddress = place.formattedAddress,
              lat = place.location?.latitude,
              lng = place.location?.longitude,
              components = place.addressComponents?.asList()?.map {
                NativeAddressComponent(
                    types = it.types, longText = it.name, shortText = it.shortName)
              } ?: emptyList())))
        }
        .addOnFailureListener { callback(Result.failure(it)) }
  }
}
